src/ReactPlaceholder.tsx

```diff
@@ -1,128 +1,142 @@
-      const mergedCustomClasses = [
-        customPlaceholder.props.className,
-        classes
-      ].filter(c => c).join(' ');
-      return React.cloneElement(customPlaceholder, { className: mergedCustomClasses });
-    } else if (customPlaceholder) {
-      return customPlaceholder;
-    }
+type RectPlaceholderProps = PlaceholderProps &
+  Omit<React.ComponentProps<typeof placeholders.rect>, 'children'> & {
+    type: 'rect';
+  };
+
+type RoundPlaceholderProps = PlaceholderProps &
+  Omit<
+    React.ComponentProps<typeof placeholders.round>,
+    'color' | 'children'
+  > & {
+    type: 'round';
+  };
 
-    const Placeholder = placeholders[type!];
+type TextPlaceholderProps = PlaceholderProps &
+  Omit<
+    React.ComponentProps<typeof placeholders.text>,
+    'color' | 'rows' | 'children'
+  > & {
+    type: 'text';
+  };
 
-    return <Placeholder {...rest} className={classes} />;
+type TextRowPlaceholderProps = PlaceholderProps &
+  Omit<
+    React.ComponentProps<typeof placeholders.textRow>,
+    'color' | 'children'
+  > & {
+    type: 'textRow';
   };
 
-  setNotReady = () => {
-    const { delay } = this.props;
+export type Props =
+  | MediaPlaceholderProps
+  | RectPlaceholderProps
+  | RoundPlaceholderProps
+  | TextRowPlaceholderProps
+  | TextPlaceholderProps
+  | CustomPlaceholderProps;
+
+const ReactPlaceholder: React.FC<Props> = ({
+  delay = 0,
+  type = 'text',
+  color = '#CDCDCD',
+  rows = 3,
+  ready: readyProp,
+  firstLaunchOnly,
+  children,
+  className,
+  showLoadingAnimation,
+  customPlaceholder,
+  ...rest
+}) => {
+  const [ready, setReady] = React.useState(readyProp);
+  const timeout = React.useRef<null | number>(null);
+
+  const getFiller = (): React.ReactElement | null => {
+    const classes = showLoadingAnimation
+      ? joinClassNames('show-loading-animation', className)
+      : className;
 
-    if (delay && delay > 0) {
-      this.timeout = window.setTimeout(() => {
-        this.setState({ ready: false });
-      }, delay);
-    } else {
-      this.setState({ ready: false });
+    if (customPlaceholder && React.isValidElement(customPlaceholder)) {
+      const mergedCustomClasses = [customPlaceholder.props.className, classes]
+        .filter(c => c)
+        .join(' ');
+      return React.cloneElement(customPlaceholder, {
+        className: mergedCustomClasses
+      });
+    } else if (customPlaceholder) {
+      return customPlaceholder;
     }
-  }
 
-  setReady = () => {
-    if (this.timeout) {
-      window.clearTimeout(this.timeout);
-    }
+    const Placeholder = placeholders[type];
+
+    return (
+      <Placeholder {...rest} color={color} rows={rows} className={classes} />
+    );
+  };
 
-    if (!this.state.ready) {
-      this.setState({ ready: true });
+  React.useEffect(() => {
+    if (!firstLaunchOnly && ready && !readyProp) {
+      if (delay && delay > 0) {
+        timeout.current = window.setTimeout(() => {
```

Comment:
this timeout should probably also be cleared before this component gets unmounted?

Reply:
Good catch!!